src/middlewared/middlewared/plugins/cloud_backup/sync.py

```diff
@@ -13,58 +16,120 @@ async def restic(middleware, job, cloud_backup, dry_run):
+                        stdin = await middleware.run_in_thread(open, zvol_name_to_path(clone), "rb")
+                    except FileNotFoundError:
+                        if i >= 5:
+                            raise
+
+                        await asyncio.sleep(1)
+                    else:
+                        break
+
+                cmd = ["--stdin", "--stdin-filename", "volume"]
+            else:
+                await check_local_path(middleware, local_path)
+
+        if cmd is None:
+            cmd = [local_path]
 
-    remote_path = get_remote_path(remote, cloud_backup["attributes"])
+        remote_path = get_remote_path(remote, cloud_backup["attributes"])
 
-    url, env = remote.get_restic_config(cloud_backup)
+        url, env = remote.get_restic_config(cloud_backup)
 
-    cmd = ["restic", "-r", f"{remote.rclone_type}:{url}/{remote_path}", "--verbose", "backup", local_path]
-    if dry_run:
-        cmd.append("-n")
+        cmd = ["restic", "-r", f"{remote.rclone_type}:{url}/{remote_path}", "--verbose", "backup"] + cmd
+        if dry_run:
+            cmd.append("-n")
 
-    env["RESTIC_PASSWORD"] = cloud_backup["password"]
+        env["RESTIC_PASSWORD"] = cloud_backup["password"]
 
-    job.middleware.logger.debug("Running %r", cmd)
-    proc = await Popen(
-        cmd,
-        env=env,
-        stdout=subprocess.PIPE,
-        stderr=subprocess.STDOUT,
-    )
-    check_progress = asyncio.ensure_future(restic_check_progress(job, proc))
-    cancelled_error = None
-    try:
+        job.middleware.logger.debug("Running %r", cmd)
+        proc = await Popen(
+            cmd,
+            env=env,
+            stdin=stdin,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+        )
+        check_progress = asyncio.ensure_future(restic_check_progress(job, proc))
+        cancelled_error = None
         try:
-            await proc.wait()
-        except asyncio.CancelledError as e:
-            cancelled_error = e
             try:
-                await middleware.call("service.terminate_process", proc.pid)
-            except CallError as e:
-                job.middleware.logger.warning(f"Error terminating restic on cloud backup abort: {e!r}")
+                await proc.wait()
+            except asyncio.CancelledError as e:
+                cancelled_error = e
+                try:
+                    await middleware.call("service.terminate_process", proc.pid)
+                except CallError as e:
+                    job.middleware.logger.warning(f"Error terminating restic on cloud backup abort: {e!r}")
+        finally:
+            await asyncio.wait_for(check_progress, None)
+
+        if cancelled_error is not None:
+            raise cancelled_error
+        if proc.returncode != 0:
+            message = "".join(job.internal_data.get("messages", []))
+            if message and proc.returncode != 1:
+                if message and not message.endswith("\n"):
```

Comment:
No reason to check `if message` again since the if branch up above already checks for this
```suggestion
                if not message.endswith("\n"):
```